README instructions for Visual Studio 2019 CMake build (#2740)
mlpack/mlpack: README.md

(or whatever directory `libmlpack.so` is installed in.)

> #### Building with Visual Studio 2019 

Reviewer: Hmm, do you think that we should make this a standalone documentation page?  It's probably good to try and keep the readme short.  Maybe we can add this to `doc/guide/build_windows.hpp` or another file in that directory?  Let me know what you think. :+1:

Author: I can certainly do that, I'll have a look at that location. But agreed, this section would be quite chunky on the `README.md`, so perhaps just a link to the new document would improve things.

Author: When I get some more time today, I'll move those over. Thanks for the review @rcurtin .

Reviewer: There are some links already to the Windows build directions:

```
There are some useful pages to consult in addition to this section:

   * Building mlpack From Source
   * Building mlpack From Source on Windows
```

If you like, we can use that second link, or perhaps we could add a third link specifically for this VS2019 integration (either linking to a separate page than the Windows build or to an anchor on the Windows build page).  Your call on what the best way is. :+1:

Author: @rcurtin I read those instructions earlier today. The instructions I provided above could be considered an alternative to the ones currently in the docs. Instead of generating the VS project with CMake, we're using VS's own CMake tools configure & generate the project.

I could incorporate the instructions above into the current docs, thus having the following sections for "opening" the project with VS:

- Generating a Visual Studio solution with CMake: this is the current instructions.
- Using VS's CMake Integration: directly work with Visual Studio's CMake integration to build the project (the instructions I added).

Also, the added OpenBLAS comments could be incorporated into the current instructions, then having the sections:

- Getting prebuilt OpenBLAS with Nuget (VS's package manager, currently in the docs).
- Building OpenBLAS with Visual Studio from source (instructions I added).

I think it's definitely best to move the changes regarded in this MR to `doc/guide/windows_build.hpp`, but I will need to incorporate these as mentioned in the points above, so it all reads smoothly :) 

I'll make these changes at some point tomorrow! Thanks again for reviewing.

Author: @rcurtin please don't resolve this discussion until I make these changes :)

Reviewer: Sounds good. :+1: I try to only resolve conversations that have been fully handled---if I resolved any too soon, feel free to unresolve them. :)

Author: @rcurtin I have incorporated these changes to the `doc/guide/build_windows.hpp` file. I haven't yet removed the ones in `README.md` for tracking purposes, but when we establish that the docs changes are satisfactory, I'll rebase and remove them :)

Author: @rcurtin let me know if any styling changes are needed as well, I might have missed something.

Reviewer: Looks great!  Thank you for taking the time to do that.  It looks like it is also in the README so maybe we should revert that change?

Author: Have reverted and cleaned up the commit history. Thanks for reviewing @rcurtin .

> This project can be build with the latest version of MS Visual Studio, given you have CMake intergration via the [C++ CMake toools for Windows](https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-160). To open the CMake project with Visual Studio, select "File->Open->CMake" in the top menu, followed by selecting the root `CMakeLists.txt` located in mlpack's root directory.

Reviewer: ```suggestion
This project can be built with the latest version of MS Visual Studio, given you have CMake integration via the [C++ CMake tools for Windows](https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-160). To open the CMake project with Visual Studio, select "File->Open->CMake" in the top menu, followed by selecting the root `CMakeLists.txt` located in mlpack's root directory.
```

Author: fixed these typos.

> In order to allow Visual Studio to configure the CMake project, the CMake configuration json will have to be editted to provide the [relevant options](#4-building-mlpack-from-source) needed to find all the dependencies. The options that you must provide to Visual Studio's CMake are:

Reviewer: ```suggestion
In order to allow Visual Studio to configure the CMake project, the CMake configuration json will have to be edited to provide the [relevant options](#4-building-mlpack-from-source) needed to find all the dependencies. The options that you must provide to Visual Studio's CMake are:
```

Author: Also fixed

* `ARMADILLO_INCLUDE_DIR`
* `ARMADILLO_LIBRARY`
* `BOOST_ROOT`
* `CEREAL_INCLUDE_DIR`
* `BLAS_LIBRARIES`
* `LAPACK_LIBRARIES`

> The CMake configuration json can be editted in Visual Studio by right clicking the root `CMakeLists.txt` in the project view, selecting **CMake settings for mlpack** and finally clicking on **edit JSON**. Adding new CMake option can be done by adding object fields with the following format to the `variables` array in the `CMakeSettings.json`:

Reviewer: ```suggestion
The CMake configuration json can be edited in Visual Studio by right clicking the root `CMakeLists.txt` in the project view, selecting **CMake settings for mlpack** and finally clicking on **edit JSON**. Adding a new CMake option can be done by adding object fields with the following format to the `variables` array in the `CMakeSettings.json`:
```

Author: Added.

```json
{
    "name": "options_name_string",
    "value": "options_value_string",
    "type" : "{BOOL|FILEPATH|PATH|STRING}"
}
```
Here is an example of a complete `CMakeSettings.json` needed to build mlpack on Visual Studio:

<details>
<summary>Click to see the example CMakeSettings.json</summary>

```json
{
  "configurations": [
    {
      "name": "x64-Debug (default)",

>       "generator": "Ninja",

Reviewer: Is Ninja generally available in MSVC 2019, or should we use the MSVC generator or something here by default?

Author: Ninja is the default generator for CMake projects in VS2019 I believe. The only thing I changed in that json file was the `variables` array.

Reviewer: Ahh, perfect, I didn't know that. :+1:  Thanks for the clarification!

      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "variables": [
        {
          "name": "ARMADILLO_INCLUDE_DIR",
          "value": "PATH/TO/CPP/DEPENDENCY/armadillo-10.1.2/include",
          "type": "PATH"
        },
        {
          "name": "ARMADILLO_LIBBRARY",
          "value": "PATH/TO/CPP/DEPENDENCY/armadillo-10.1.2/lib/armadillo.lib",
          "type": "PATH"
        },
        {
          "name": "CEREAL_INCLUDE_DIR",
          "value": "PATH/TO/CPP/DEPENDENCY/cereal-1.3.0/include",
          "type": "PATH"
        },
        {
          "name": "BUILD_ROOT",
          "value": "PATH/TO/CPP/DEPENDENCY/boost_1_66_0",
          "type": "PATH"
        },
        {
          "name": "BOOST_INCLUDEDIR",
          "value": "PATH/TO/CPP/DEPENDENCY/boost_1_66_0",
          "type": "PATH"
        },
        {
          "name": "BLAS_LIBRARIES",
          "value": "PATH/TO/CPP/DEPENDENCY/OpenBLAS/lib/openblas.lib",
          "type": "PATH"
        },
        {
          "name": "LAPACK_LIBRARIES",
          "value": "PATH/TO/CPP/DEPENDENCY/OpenBLAS/lib/openblas.lib",
          "type": "PATH"
        }
      ]
    }
  ]
}
```
</details>

> ##### Blas and LAPACK on Windows

Reviewer: ```suggestion
##### BLAS and LAPACK on Windows
```
Technically BLAS is an acronym too, just like LAPACK. :)

Author: You are correct. I have replaced all the instances of `Blas` with `BLAS`

> Unfortunately, the support for building LAPACK and Blas on Windows is quite poor, due to the need for Fortran compiler and libraries. The easiest method to get the necessary Blas/LAPACK libraries is to compile OpenBLAS with LLVM's `clang-cl` and `flang` to produce the required static library (`.lib`) files compatible with the MSVC compiler. A comprehensive guide on the [compilation of OpenBLAS for Windows can be found here](https://github.com/xianyi/OpenBLAS/wiki/How-to-use-OpenBLAS-in-Microsoft-Visual-Studio).

Reviewer: ```suggestion
Unfortunately, the support for building LAPACK and BLAS on Windows is quite poor, due to the need for Fortran compiler and libraries. The easiest method to get the necessary BLAS/LAPACK libraries is to compile OpenBLAS with LLVM's `clang-cl` and `flang` to produce the required static library (`.lib`) files compatible with the MSVC compiler. A comprehensive guide on the [compilation of OpenBLAS for Windows can be found here](https://github.com/xianyi/OpenBLAS/wiki/How-to-use-OpenBLAS-in-Microsoft-Visual-Studio).
```

Author: replaced.

> One could always download prebuilt `LAPACK` and `Blas` libraries for Windows. However, there are few official sources, and some of those libraries may require further `dll`s at runtime which may not be available in your system.

Reviewer: ```suggestion
One could always download prebuilt `LAPACK` and `BLAS` libraries for Windows. However, there are few official sources, and some of those libraries may require further `dll`s at runtime which may not be available in your system.
```

Author: replaced.

> It you choose to build `OpenBLAS` from source, make sure that `LAPACK` functions are also built. Finally, make sure that the `openblas.lib` library is linked in your 	`Armadillo` build, as well as the library path used for the CMake options `BLAS_LIBRARIES` `LAPACK_LIBRARIES` in the mlpack CMake project.

Reviewer: ```suggestion
It you choose to build `OpenBLAS` from source, make sure that `LAPACK` functions are also built. Finally, make sure that the `openblas.lib` library is linked in your `Armadillo` build, as well as the library path used for the CMake options `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` in the mlpack CMake project.
```

Author: Also added suggestions.
